### PR TITLE
[FEATURE] add FWHM to PP-HiRes and MassTrace

### PR DESCRIPTION
**Pull request by cbielow**
- PP-HiRes can now optionally report the FWHM of mz-peaks (disabled by default)
 - MassTraceExtractor can do the same; as a meta-value in the featureXML

Minor changes:
 - minimum number of raw peaks for picking is now 3 (as documented) and not 4
 - minor fixes and code-rearrangements

Note: @hroest; I had to duplicate the TOPP test output since the lowmemory option produces a different "dataProcessing" header in mzML since each floatDataArray requires its own section (seems that mzML designers decided this would be a good idea...) and the writer cannot possibly know the number of floatDataArrays of future spectra in advance.  

An upcoming PR will make use of this in FF-Metabo (and fix a couple of other things there)

### CODE REVIEW

**Comment by mwalzer** [comment]
apparently, our structure is like that each fp has its own dp object and I guess that was modelled so that every spectrum could, in theory, be treated differently and documented so...
but you can have more than one processing method in a dp by schema, so that is up to the bundling logic of the writer.

**Comment by cbielow** [comment]
tests pass (failed tests are unrelated). 
Merge?

File: src/openms/include/OpenMS/TRANSFORMATIONS/RAW2PEAK/PeakPickerHiRes.h [hunk 1]
```diff
@@ -342,10 +348,62 @@ namespace OpenMS
...
+                }
+              } while(fabs(int_mid - fwhm_int) > threshold);
+            }
+            const double fwhm_left_mz = mz_mid;
+
+            // right ...
+            double mz_right = peak_raw_data.rbegin()->first;
+            mz_center = max_peak_mz;
+            if (peak_spline.eval(mz_right) > fwhm_int)
+            { // the spline ends before half max is reached -- take the leftmost point (probably an underestimation)
```
Comment (hroest) [inline src/openms/include/OpenMS/TRANSFORMATIONS/RAW2PEAK/PeakPickerHiRes.h]: leftmost -> rightmost point?

File: src/openms/include/OpenMS/TRANSFORMATIONS/RAW2PEAK/PeakPickerHiRes.h [hunk 2]
```diff
@@ -342,10 +348,62 @@ namespace OpenMS
...
+              mz_mid = mz_left;
+            } else
+            {
+              do 
+              {
+                mz_mid = mz_left / 2 + mz_center / 2;
+                int_mid = peak_spline.eval(mz_mid);
+                if (int_mid < fwhm_int) {
+                  mz_left = mz_mid;
```
Comment (hroest) [inline src/openms/include/OpenMS/TRANSFORMATIONS/RAW2PEAK/PeakPickerHiRes.h]: can you change this to OpenMS style:

```
if (int_mid < fwhm_int) 
{
  mz_left = mz_mid;
}
```

**Comment by hroest** [comment]
I have a few more comments, see text. 

Another issue is whether we want to store ppm for FWHM - this makes only sense in the `m/z` dimension but if we do chromatographic peak picking then it does not make sense. The `PeakPickerMRM` and also the metabolic peakpicker do chromatographic peak picking and in that case it does not make sense to report FWHM in ppm. either we should add a parameter for ppm / no-ppm or have the client compute ppm from the data

**Comment by cbielow** [comment]
Good point. I've extended the params, such that the user can choose what he'd like to see now.
It's hard to automatically decide if ppm or absolute FWHM should be chosen, since the pick() function only works on spectra, to which also Chromatograms are temporarily converted. So I'd have to introduce a new function parameter for a lot of functions and alias functions which I consider overkill.
If Spectrum and Chromatogram would have the same interface (e.g. ::Settings and not ::SpectrumSettings vs. ChromatogramSettings) one could write a generic pick() template, and also save a lot of internal copying.... but that's maybe not for today.

**Comment by cbielow** [comment]
failed checks are unrelated

**Comment by timosachsenberg** [comment]
+1

**Pull request merged**